### ISSUE TITLE
Available SSL/TLS Protocols

### ISSUE THREAD

**Issue by MSP-Greg** [eventmachine/eventmachine#857]
I'm working with OpenSSL 1.1.1, and while investigating what needs to be updated, running some test issues.

As best I can tell, EM has no way to easily query what protocols are available.  I haven't built 1.0.2 for quite a while, but both 1.1.0 & 1.1.1 allow configuration of available protocols, in particular, removing SSL2 & SSL3.

EM has the constants `EM_PROTO_*`.  Could these be undefined if a protocol is not available thru OpenSSL?

Thanks, Greg

**Comment by sodabrew** [comment]
The variables `EM_PROTO_*` are internal to EM and always available even if the underlying SSL library does not provide that protocol: https://github.com/eventmachine/eventmachine/blob/b430a52b1e5166d1d1f5d388cb1f92a4f15b7bf1/ext/eventmachine.h#L42

**Comment by MSP-Greg** [comment]
Another API thought is that both OpenSSL & Ruby OpenSSL are supporting min & max, such that
```ruby
start_tls(:ssl_version => %w(TLSv1_2))
```
Could instead be:
```ruby
start_tls(:min_version => %w(TLSv1_2))
```

**Comment by sodabrew** [comment]
I specifically sidestepped the min version, instead the ssl_version array is a whitelist of what's allowed. But maybe that turns out to be an incongruent API?

**Comment by MSP-Greg** [comment]
I suspect this is the kind of issue where no matter what changes might be made, someone will ask for it to work differently.

Personally, I think the min version would be helpful, but I'm not sure if I can write the code for it.  Remember, I'm a 'light c' person.

The ssl_version array is fine, but it doesn't match up well with the current OpenSSL & Ruby OpenSSL API's.  I'm familiar with Ruby & Puma SSL tests, and I've never seen a test that used a non-contiguous set of protocols.

People may want SSLv3 for internal/legacy use, and what version of OpenSSL EM is built with may be out of their control.

So, I've added two constants `OPENSSL_NO_SSL2` and `OPENSSL_NO_SSL3` which will be true if the respective protocol is disabled.  About to run a build & test with OpenSSL 1.0.1 thru 1.1.1 and see if they work correctly.  These hopefully will allow testing to truly know what to test without making assumptions about what protocols are available in what OpenSSL version...

When and if it works locally, I'll post the patch for it before doing a PR...

**Comment by MSP-Greg** [comment]
@sodabrew

The constants (`OPENSSL_NO_SSL2` and `OPENSSL_NO_SSL3`) seem to be correct, but now I have a few errors in test_resolver.rb in Ruby 2.0, 2.1, & 2.3.  Re the constants, they're determined compile time; maybe they should be runtime?  I could see someone building EM with SSL3 enabled in OpenSSL, then updating to a new OpenSSL 'alpha' version with it disabled?

Attached is a patch with a txt extension added.  The patch for `em_test_helper.rb` are just temporary.

[em_TLSv13_noSSL3.patch.txt](https://github.com/eventmachine/eventmachine/files/2280431/em_TLSv13_noSSL3.patch.txt)

**Comment by MSP-Greg** [comment]
Guess I ended up working on EM a bit more than expected.

Just to check status, I piled all the current changes I have into a branch, and hoped to run a test on Travis, but of course, it won't let me...  That's one thing Appveyor has together that Travis doesn't.

Anyway, the commit is [here](https://github.com/MSP-Greg/eventmachine/commit/3a8b30ba0e6812ab5828e49cb74105e33a078d35).  There are three lines in `tests/em_test_helper.rb` (lines 15-17) that are just debug lines for the new constants.  they appear to be working correctly.

The test logs 7z file for 64 bit is [here](https://ci.appveyor.com/api/buildjobs/o6c0uf0vf09shr22/artifacts/test_logs-x64-mingw32.7z), the 32 bit 7z is [here](https://ci.appveyor.com/api/buildjobs/iwxd3648a2i6jhwq/artifacts/test_logs-x86-mingw32.7z).  All tests passed from 2.0 thru trunk, both 32 & 64.

Testing with 1.1.1 locally, the only failure is test_ssl_extensions.rb:43:in test_tlsext_sni_hostname.

In particular, note the changes in the c files, and also [`test_ssl_protocols.rb`](https://github.com/MSP-Greg/eventmachine/blob/3a8b30ba0e6812ab5828e49cb74105e33a078d35/tests/test_ssl_protocols.rb).


Re `test_ssl_protocols.rb`, `SSL_AVAIL` is now no longer dependent on any version info.  In trying to add some tests for TLSv1.3, I decided that maybe it was also time to consolidate the common code.  I also omitted the global variables, but accidentally left two lines in the code (119 & 124).

**Comment by xnox** [comment]
I haven't looked, but for sni to work with tls1.3 one needs ensure the client context validates hostname. E.g. there have been fixes like that elsewhere. An example I have at hand, is a python one, but hopefully simple enough to understand https://sources.debian.org/src/python-boto/2.44.0-1.1/debian/patches-for-reference/boto-try-to-add-SNI-support-v2.patch/

**Comment by MSP-Greg** [comment]
Travis is currently testing with OpenSSL 1.0.2 & 1.1.0.  Future Appveyor will use OpenSSL 1.1.1.  The sni issue was resolved in OpenSSL 1.1.1, but the API may change in 1.2.
